libcudacxx/.upstream-tests/test/cuda/barrier/cp_async_bulk_tensor_generic.h

```diff
@@ -0,0 +1,270 @@
+
+    // Check smem
+    for (int i = threadIdx.x; i < smem_len; i += blockDim.x) {
+        int gmem_lin_idx = smem_lin_idx_to_gmem_lin_idx(i, smem_coord, smem_dims, gmem_dims);
+        assert(smem_buffer[i] == gmem_lin_idx);
+    }
+
+    __syncthreads();
+
+    // Update smem
+    for (int i = threadIdx.x; i < smem_len; i += blockDim.x) {
+        smem_buffer[i] = 2 * smem_buffer[i] + 1;
+    }
+    cde::fence_proxy_async_shared_cta();
+    __syncthreads();
+
+    // Write back to global memory:
+    if (threadIdx.x == 0) {
+        cp_tensor_shared_to_global(global_tensor_map, smem_coord, smem_buffer);
+        cde::cp_async_bulk_commit_group();
+        cde::cp_async_bulk_wait_group_read<0>();
+    }
+    __threadfence();
+    __syncthreads();
+
+    // // TEAR-DOWN: check that global memory is correct
+    for (int i = threadIdx.x; i < smem_len; i += blockDim.x) {
+        int gmem_lin_idx = smem_lin_idx_to_gmem_lin_idx(i, smem_coord, smem_dims, gmem_dims);
+
+        assert(gmem_tensor[gmem_lin_idx] == 2 * gmem_lin_idx + 1);
+    }
+    __syncthreads();
+}
+
+#ifndef TEST_COMPILER_NVRTC
+PFN_cuTensorMapEncodeTiled get_cuTensorMapEncodeTiled() {
+    void* driver_ptr = nullptr;
+    cudaDriverEntryPointQueryResult driver_status;
+    auto code = cudaGetDriverEntryPoint("cuTensorMapEncodeTiled", &driver_ptr, cudaEnableDefault, &driver_status);
+    assert(code == cudaSuccess && "Could not get driver API");
+    return reinterpret_cast<PFN_cuTensorMapEncodeTiled>(driver_ptr);
+}
+#endif
+
+#ifndef TEST_COMPILER_NVRTC
+template <typename T>
+CUtensorMap map_encode(T *tensor_ptr, std::initializer_list<int> gmem_dims, std::initializer_list<int> smem_dims) {
+    // https://docs.nvidia.com/cuda/cuda-driver-api/group__CUDA__TENSOR__MEMORY.html
+    CUtensorMap tensor_map{};
+    assert(gmem_dims.size() == smem_dims.size());
+    // rank is the number of dimensions of the array.
+    int rank = gmem_dims.size();
+    uint64_t size[rank];
+    for (int i  = 0; i < rank; ++i) {
+        size[i] = gmem_dims.begin()[i];
+    }
+    // The stride is the number of bytes to traverse from the first element of one row to the next.
+    // It must be a multiple of 16.
+    uint64_t stride[rank - 1]; // = {GMEM_WIDTH * sizeof(int)};
+    int base_stride = sizeof(T);
+    for (int i = 0; i < rank - 1; ++i) {
+        base_stride *= gmem_dims.begin()[i];
+        stride[i] = base_stride;
+    }
+    // The box_size is the size of the shared memory buffer that is used as the
+    // destination of a TMA transfer.
+    uint32_t box_size[rank]; // = {SMEM_WIDTH, SMEM_HEIGHT};
+    for (int i = 0; i < rank; ++i) {
+        box_size[i] = smem_dims.begin()[i];
+    }
+    // The distance between elements in units of sizeof(element). A stride of 2
+    // can be used to load only the real component of a complex-valued tensor, for instance.
+    uint32_t elem_stride[rank]; // = {1, 1};
+    for (int i = 0; i < rank; ++i) {
+        elem_stride[i] = 1;
+    }
+
```

Comment:
Why not use the initializer?

Reply:
`rank` is a runtime value. I am not sure how to initialize an initializer list with a variable number of ones. I updated the comment to clarify.

Reply:
For obvious reasons `initializer_list::size()` is not constexpr in c++11 so we cannot use it.

Thanks for the reminder